adapters/Vungle/VungleAdapter/GADMAdapterVungleRouter.m

```diff
@@ -126,67 +126,87 @@ - (BOOL)addDelegate:(nonnull id<GADMAdapterVungleDelegate>)delegate {
+        GADMAdapterVungleBannerRequest *vungleBannerRequest = nil;
+        while (vungleBannerRequest = [enumerator nextObject]) {
+          // There is already a banner delegate with same placementID
+          // but different uniquePubRequestID in _bannerDelegates.
+          if ([vungleBannerRequest.placementID isEqualToString:delegate.bannerRequest.placementID]) {
             return NO;
-          } else {
-            NSOrderedSet *set = [_bannerDelegates objectForKey:bannerRequest];
-            for(id<GADMAdapterVungleDelegate> delegate in set) {
-              if (delegate.bannerState == BannerRouterDelegateStateWillPlay || delegate.bannerState == BannerRouterDelegateStatePlaying) {
-                return NO;
-              }
+          } else if (vungleBannerRequest.placementID.length > 0 && ![vungleBannerRequest.placementID isEqualToString:delegate.bannerRequest.placementID]) {
+            // There is already a banner delegate with different placementID in _bannerDelegates.
+            if (!_bannerRequest) {
+              _bannerRequest = [vungleBannerRequest copy];
             }
+            return NO;
           }
         }
 
-        NSMutableOrderedSet *set = [[NSMutableOrderedSet alloc] initWithObject:delegate];
-        GADMAdapterVungleMapTableSetObjectForKey(_bannerDelegates, delegate.bannerRequest, set);
-        _bannerPlacementID = [delegate.desiredPlacement copy];
-      } else if (delegate && [_bannerDelegates objectForKey:delegate.bannerRequest]) {
-        NSMutableOrderedSet *set = [[_bannerDelegates objectForKey:delegate.bannerRequest] mutableCopy];
-        if (![set containsObject:delegate]) {
-          GADMAdapterVungleMapTableRemoveObjectForKey(_bannerDelegates, delegate.bannerRequest);
-          [set addObject:delegate];
-          GADMAdapterVungleMapTableSetObjectForKey(_bannerDelegates, delegate.bannerRequest, set);
-          _bannerPlacementID = [delegate.desiredPlacement copy];
-        }
+        GADMAdapterVungleMapTableSetObjectForKey(_bannerDelegates, delegate.bannerRequest, delegate);
+        _bannerRequest = [delegate.bannerRequest copy];
+      } else {
+        id<GADMAdapterVungleDelegate> bannerDelegate = [_bannerDelegates objectForKey:delegate.bannerRequest];
+
+        /* The isRequestingBannerAdForRefresh flag is used for an edge case, when the old Banner
+         * delegate is removed from _bannerDelegates and there is a refresh Banner delegate doesn't
+         * construct Banner view successfully and add to _bannerDelegates yet. Adapter cannot set
+         * _bannerRequest to nil to prevent requesting another Banner ad with different bannerRequest.
+         */
+        bannerDelegate.isRequestingBannerAdForRefresh = YES;
+        delegate.isRefreshedForBannerAd = YES;
       }
     }
   }
   return YES;
 }
 
+- (void)replaceOldBannerDelegate:(nonnull id<GADMAdapterVungleDelegate>)delegate {
```

Comment:
nit: Suggest renaming to:

replaceOldBannerDelegateWithDelegate:

since what's getting passed in is not the old banner delegate but actually the new one :)

Reply:
@ericleich Got it, I will change the name to `replaceOldBannerDelegateWithDelegate:`. Thanks a lot for the guidance. ^_^

Reply:
@ericleich Done, thanks a lot. ^_^